### ISSUE TITLE
[from customer] nios_txt_record module cannot handle multiple TXT records

### ISSUE THREAD

**Issue by sugitk** [ansible/ansible#62377]
##### SUMMARY

nios_txt_record module cannot handle multiple TXT records

##### ISSUE TYPE
- Bug Report

##### ANSIBLE VERSION

```paste below
2.8.2
```

##### CONFIGURATION

```paste below

```

##### STEPS TO REPRODUCE

Add the TXT record.  Change the "text" field value and add it again.  Instead of adding a second record with a different value, it replaces the first record.

I think it is an expected behavior with an idempotent way, but there is no way I can find to add multiple TXT records.

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sugitk** [comment]
The customer complained that it should be checked the `view`, `name`, and `text` parameters for uniqueness.  I agree with his opinion.

**Comment by sivel** [comment]
!component =lib/ansible/modules/net_tools/nios/nios_txt_record.py

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/net_tools/nios/nios_txt_record.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/net_tools/nios/nios_txt_record.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @NilashishC @Qalthos @coreywan @danielmellado @ganeshrn @justjais @sganesh-infoblox @trishnaguha
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by WRJFontenot** [comment]
I am the customer that opened two support cases with RedHat about the nios_txt_record module:

CASE 02473253: Ansible module nios_txt_record won't add TXT records if multiple TXT records already exist.
CASE 02473244: Ansible nios_txt_record module replaces existing TXT records instead of adding additional records.

Both issues are likely caused by the same problem.  In the documentation for [nios_txt_record](https://docs.ansible.com/ansible/latest/modules/nios_txt_record_module.html#nios-txt-record-module), the view and name parameters are required, but the text parameter is not.  This is wrong. 
The TXT record type is just like the A record type, there can be many records with the same name but different values.  For this reason, in the nios_a_record module, the ipaddress parameter is a required field.  

The text parameter in nios_txt_record should also be a required field and the nios_txt_record module should be looking at the combination of the view:name:text parameters values to determine if an operation is idempotent.

If I run the following playbook:

`---

- hosts: example_grid_primary
  connection: local

  tasks:
    - name: Add/Delete TXT Records
      nios_txt_record:
        provider: "{{nios_provider}}"
        comment: CHG0000000
        view: External
        name: _example.example.com
        text: thequickbrownfoxjumpedoverthelazydog
        state: present

    - name: Add/Delete TXT Records
      nios_txt_record:
        provider: "{{nios_provider}}"
        comment: CHG0000000
        view: External
        name: _example.example.com
        text: fourscoreandsevenyearsagoourfathers
        state: present

...
`
I should get this as my response to a TXT query for _example.example.com:

_example.example.com.	900	IN	TXT	"thequickbrownfoxjumpedoverthelazydog"
_example.example.com.	900	IN	TXT	"fourscoreandsevenyearsagoourfathers"

Instead I get:

_example.example.com.	900	IN	TXT	"fourscoreandsevenyearsagoourfathers"

The above example only works if there isn't already a TXT record with the name "_example.example.com."

If there are already multiple TXT records with the same name because they were imported or added via the Infoblox admin interface, then when you try to add another TXT record with the same name using the nios_txt_record module, you get an error:

[add_txt_records_errors.txt](https://github.com/ansible/ansible/files/3628087/add_txt_records_errors.txt)

**Comment by WRJFontenot** [comment]
When deciding priority for bug fixes, keep in mind  if a customer today operating in only one AWS region uses the nios_txt_module to add a second TXT record for AWS SES so they can operate in a second region, their first SES TXT record will be replaced the second and they will experience an email failure with all their email in their original AWS region.

We only avoided this fate because we already operate in multiple AWS regions and had multiple TXT records with the same name, so this bug's second failure mode actually saved us from having a Severity 1 outage.

**Comment by justjais** [comment]
@s-giri @krisvasudevan can you guys plz look into the above issue and update with your findings. It's a customer request, it would be great if we can come with resolution asap.

**Comment by sugitk** [comment]
@s-giri @krisvasudevan 
Any update?  We appreciate it if you could handle this issue which the customer is now facing.

**Comment by IshwarKanse** [comment]
@s-giri @krisvasudevan @justjais  Anybody working on this ? Its a customer request and this has been opened for almost a month.

**Comment by abenokraitis** [comment]
This has been escalated to Infoblox as they are a Certified Ansible Partner, and should be ultimately responsible and accountable for the fix. Ansible will do the needful to include in the applicable releases once a suitable fix is provided.

**Comment by justjais** [comment]
@sugitk @IshwarKanse can you ask the respective customer to raise the issue with reference to this issue @infoblox side as well

**Comment by IshwarKanse** [comment]
@justjais I've opened an issue with Infoblox on behalf of the customer.

**Comment by IshwarKanse** [comment]
Following is the link to this issue raised on Infoblox's community forum as it was requested by their support team https://community.infoblox.com/t5/Network-Change-Configuration/Issues-with-Ansible-nios-txt-record-Infoblox-module/m-p/18688#M2881 I've also asked the customer to open a case with them.

**Comment by Akasurde** [comment]
@justjais Can you please confirm if this issue is fixed via #65639? Thanks,

needs_info

**Comment by justjais** [comment]
@sugitk can you please verify the fix from `devel` and update the issue. Thanks

**Comment by krisvasudevan** [comment]
The fix has been added to a new pull request #66014

**Comment by tadeboro** [comment]
We just tested this locally using Infoblox 8.4.4 and Ansible from the devel branch and the issue is fixed there. The PR that introduced the fix is #65369. The stable-2.9 branch is still broken though.

**Comment by krisvasudevan** [comment]
The PR that introduced the fix is actually #66014.

**Comment by tadeboro** [comment]
> The PR that introduced the fix is actually #66014.

No, because that PR is not merged yet and the stings are already working.

**Comment by tadeboro** [comment]
I created two minimalistic bugfixes (#67120 and #67121) for Ansible 2.8 and 2.9. They both contain the same single-line fix and should be easy to review.

**Comment by ansibot** [comment]
@sugitk This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by WRJFontenot** [comment]
I am the customer waiting on this fix.  Do not close this issue until it is resolved.

**Comment by tadeboro** [comment]
As I said, I tested the fixes for:

 1. the `devel` branch: #65369,
 2. the `stable-2.9` branch: #67120,
 3. the `stable-2.8` branch: #67121

The playbook I used to test this is the following:

```
---
- name: Testing Infoblox TXT records
  hosts: localhost
  gather_facts: false

  tasks:
    - name: Add a first TXT record
      nios_txt_record:
        name: bla.tadej.sem
        text: text1
        provider: &prov
          host: 172.16.93.148
          username: admin
          password: infoblox

    - name: Add a second TXT record
      nios_txt_record:
        name: bla.tadej.sem
        text: text2
        provider: *prov
```

If I run this playbook using the unpatched Ansible 2.8 or 2.9, I get the following result in the Infoblox dashboard:

![ib1](https://user-images.githubusercontent.com/70951/74163057-1568a500-4c22-11ea-9936-63d50677a03c.png)

It can be clearly seen from the screenshot above that the second task replaced the TXT record that the first task created. Running the same playbook using patched Ansible (2.8, 2.9 and devel version) produces the following result:

![ib2](https://user-images.githubusercontent.com/70951/74163203-5365c900-4c22-11ea-8173-984e8aaa5897.png)

Hopefully, this confirms that the fixes actually work. As stated before, the development version is already patched (so Ansible 2.10 will be free of this bug), and there are two minor fixes for versions 2.8 and 2.9 if anyone is interested in having those two versions fixed.

Is there anything else needed here?

**Comment by gberginc** [comment]
@WRJFontenot @sugitk Could you check the scenario tested by @tadeboro and the results? PR #67120 for Ansible 2.9 was merged earlier today and should be shipped with the next release of Ansible 2.9.

**Comment by WRJFontenot** [comment]
I'll test it as soon as the next release is available on EPEL.

**Comment by WRJFontenot** [comment]
> As I said, I tested the fixes for:
> 
> 1. the `devel` branch: #65369,
> 2. the `stable-2.9` branch: #67120,
> 3. the `stable-2.8` branch: #67121
> 
> The playbook I used to test this is the following:
> 
<SNIP>

> Is there anything else needed here?
<SNIP>

Does the update TXT module work if there are already multiple TXT records with the same name?  The old module would crash if there was more than one existing TXT record with the same name.  Your first test started with 0 records, the second with 1 record.  You should do a third to see if it works with more than one TXT record with the same name.

Also, can you remove a given TXT record without impacting the remaining records?

**Comment by tadeboro** [comment]
> Does the update TXT module work if there are already multiple TXT records with the same name? The old module would crash if there was more than one existing TXT record with the same name. Your first test started with 0 records, the second with 1 record. You should do a third to see if it works with more than one TXT record with the same name.
>
> Also, can you remove a given TXT record without impacting the remaining records?

Maybe I simplified the use-case a bit too much. I did test those scenarios while preparing PRs for Ansible 2.8 and 2.9, and all of the things you are asking about here are solved in the `devel` and `stable-2.9` branches.

The module crashed previously because it always excepted that the backend has at most one matching entry. Now that the `text` field is part of the parameters that are used to look up the existing entries, the at-most-one invariant always holds, and the crash cannot happen anymore. This same line of reasoning also explains why it is now possible to remove only one specific TXT record without touching anything else.

And while I am replying, Ansible 2.8 may not get fixed because it is currently accepting only critical bug fixes. So it is up to the release manager to decide if this issue is critical enough to warrant a fix in 2.8.

**Comment by WRJFontenot** [comment]
I confirm this problem is fixed in Ansible 2.9.5.  I can add and delete multiple TXT records with the same name even when other TXT records with the same name are already present.  Good work!
I will close the underlying RedHat support case so @sugitk can close this bug.

**Comment by sugitk** [comment]
@WRJFontenot 
Thank you very much for your report and a long team patience.  I will close this issue out.
